File: packages/typescript/src/aat/pop.ts
/**
 * Proof-of-possession verification for AAT -01.
 *
 * PoP binds a presentation to the leaf holder's private key and to the exact
 * argument map of one invocation. Without it, a chain is only evidence that
 * authority was issued — not that the presenter holds it.
 */

import canonicalize from 'canonicalize';
import { compactVerify, decodeProtectedHeader, importJWK } from 'jose';

import { denial, type Denial } from '../reasons.js';
import type { PopPayload, ToolArguments } from './claims.js';
import type { VerifiedDelegationChain } from './verify.js';

const PERMITTED_ALGORITHMS: ReadonlySet<string> = new Set(['EdDSA']);

/** Clock tolerance for a proof of possession, in seconds. */
export const POP_FRESHNESS_SECONDS = 300;

export type PopResult = { ok: true } | { ok: false; denials: Denial[] };

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function looksLikePopPayload(value: unknown): value is PopPayload {
  if (!isPlainObject(value)) return false;
  return (
    typeof value['jti'] === 'string' &&
    typeof value['iat'] === 'number' &&
    typeof value['aat_id'] === 'string' &&
    typeof value['aat_tool'] === 'string' &&
    isPlainObject(value['hta'])
  );
}

/**
 * Verify a PoP JWT against a verified chain, a tool, and an argument map.
 *
 * Argument binding is compared as JCS-canonical bytes, so key order and
 * insignificant formatting differences do not matter while any change of value
 * does.
 */
export async function verifyProofOfPossession(
  pop: string,
  chain: VerifiedDelegationChain,
  tool: string,
  args: ToolArguments,
  now: number = Math.floor(Date.now() / 1000),
): Promise<PopResult> {
  if (typeof pop !== 'string' || pop.length === 0) {
    return {
      ok: false,
      denials: [denial('pop_missing', 'pop', 'No proof of possession was presented.')],
    };
  }

  let header: { alg?: string };
  try {
    header = decodeProtectedHeader(pop);
  } catch {
    return {
      ok: false,
      denials: [denial('pop_malformed', 'pop', 'Proof-of-possession header is not valid.')],
    };
  }

  const alg = header.alg;
  if (typeof alg !== 'string' || !PERMITTED_ALGORITHMS.has(alg)) {
    return {
      ok: false,
      denials: [
        denial(
          'algorithm_not_permitted',
          'pop',
          `Proof-of-possession algorithm ${JSON.stringify(alg ?? null)} is not permitted; expected EdDSA.`,
        ),
      ],
    };
  }

  const leaf = chain.tokens[chain.tokens.length - 1];
  if (leaf === undefined) {
    return { ok: false, denials: [denial('chain_empty', 'pop', 'Chain has no leaf token.')] };
  }

  const jwk = leaf.payload.cnf.jwk;
  if (jwk['kty'] !== 'OKP') {
    return {
      ok: false,
      denials: [denial('holder_key_invalid', 'pop', 'Leaf holder key is not an OKP key.')],
    };
  }

  let payload: unknown;
  try {
    const key = await importJWK(jwk, alg);
    const verified = await compactVerify(pop, key);
    payload = JSON.parse(new TextDecoder().decode(verified.payload)) as unknown;
  } catch {
    return {
      ok: false,
      denials: [
        denial(
          'pop_signature_invalid',
          'pop',
          'Proof of possession is not signed by the leaf holder key.',
        ),
      ],
    };
  }

  if (!looksLikePopPayload(payload)) {
    return {
      ok: false,
      denials: [denial('pop_malformed', 'pop', 'Proof of possession is missing required claims.')],
    };
  }

  const denials: Denial[] = [];

  if (payload.aat_id !== leaf.payload.jti) {
    denials.push(
      denial('pop_token_mismatch', 'pop', 'Proof of possession names a different leaf token.'),
    );
  }
  if (payload.aat_tool !== tool) {
    denials.push(
      denial('pop_tool_mismatch', 'pop', 'Proof of possession names a different tool.', { tool }),
    );
  }
  // Step 7g: a proof is bound to one invocation, so an old one must not be
  // replayable indefinitely.
  if (Math.abs(payload.iat - now) > POP_FRESHNESS_SECONDS) {
    denials.push(
      denial('pop_stale', 'pop', 'Proof of possession is outside the freshness window.'),
    );
  }

  if (canonicalize(payload.hta) !== canonicalize(args)) {
    denials.push(
      denial(
        'pop_binding_mismatch',
        'pop',
        'Proof of possession is bound to a different argument map.',
        { tool },
      ),
    );
  }

  return denials.length > 0 ? { ok: false, denials } : { ok: true };
}

File: python/oaaf/pop.py
"""Proof-of-possession verification (AAT -01).

Binds a presentation to the leaf holder's key and to the exact argument map of
one invocation. Argument binding is compared as RFC 8785 canonical bytes.
"""

from __future__ import annotations

from typing import Optional

from . import crypto
from .reasons import Denial, denial
from .verify import VerifiedChain

POP_FRESHNESS_SECONDS = 300


def verify_pop(
    pop: str,
    chain: VerifiedChain,
    tool: str,
    args: dict,
    now: int,
) -> list[Denial]:
    if not isinstance(pop, str) or len(pop) == 0:
        return [denial("pop_missing", "pop", "No proof of possession was presented.")]
    if not crypto.alg_permitted(pop):
        return [denial("algorithm_not_permitted", "pop", "Proof-of-possession algorithm is not permitted.")]

    leaf = chain.tokens[-1]
    jwk = leaf["cnf"]["jwk"]
    if jwk.get("kty") != "OKP":
        return [denial("holder_key_invalid", "pop", "Leaf holder key is not an OKP key.")]

    payload = crypto.verify_compact_jws(pop, jwk)
    if payload is None:
        return [denial("pop_signature_invalid", "pop", "Proof of possession is not signed by the leaf holder key.")]
    if not _looks_like_pop(payload):
        return [denial("pop_malformed", "pop", "Proof of possession is missing required claims.")]

    out: list[Denial] = []
    if payload["aat_id"] != leaf["jti"]:
        out.append(denial("pop_token_mismatch", "pop", "Proof of possession names a different leaf token."))
    if payload["aat_tool"] != tool:
        out.append(denial("pop_tool_mismatch", "pop", "Proof of possession names a different tool.", tool=tool))
    if abs(payload["iat"] - now) > POP_FRESHNESS_SECONDS:
        out.append(denial("pop_stale", "pop", "Proof of possession is outside the freshness window."))
    if crypto.jcs(payload["hta"]) != crypto.jcs(args):
        out.append(denial("pop_binding_mismatch", "pop", "Proof of possession is bound to a different argument map.", tool=tool))
    return out


def pop_audience(pop: str) -> Optional[str]:
    parts = pop.split(".")
    if len(parts) != 3:
        return None
    raw = crypto.b64url_decode(parts[1])
    if raw is None:
        return None
    try:
        import json

        payload = json.loads(raw)
    except ValueError:
        return None
    aud = payload.get("aat_aud") if isinstance(payload, dict) else None
    return aud if isinstance(aud, str) else None


def _looks_like_pop(v: dict) -> bool:
    return (
        isinstance(v.get("jti"), str)
        and isinstance(v.get("iat"), (int, float))
        and isinstance(v.get("aat_id"), str)
        and isinstance(v.get("aat_tool"), str)
        and isinstance(v.get("hta"), dict)
    )
